### ISSUE TITLE
BadStatusLine errros

### ISSUE THREAD

**Issue by jorilallo** [pusher/pusher-http-python#23]
Recently I started getting `BadStatusLine` errors. Here's the traceback:

```
File "/app/inc/items/tasks.py", line 19, in live_item_update
pusher['private-%s' % member.pk].trigger('items:%s:change' % item_pk, {})
File "/app/.heroku/venv/lib/python2.7/site-packages/pusher/__init__.py", line 68, in trigger
status, resp_content = self.send_request(signed_path, json_data, timeout=timeout)
File "/app/.heroku/venv/lib/python2.7/site-packages/pusher/__init__.py", line 98, in send_request
resp = http.getresponse()
File "/usr/local/lib/python2.7/httplib.py", line 1027, in getresponse
response.begin()
File "/usr/local/lib/python2.7/httplib.py", line 407, in begin
version, status, reason = self._read_status()
File "/usr/local/lib/python2.7/httplib.py", line 371, in _read_status
raise BadStatusLine(line)
BadStatusLine: ''
```

**Comment by zimbatm** [comment]
1.0 is now using requests as a backend by default which should be much more reliable. Feel free to re-open if you're seeing the issue again obviously.

**Comment by amackera** [comment]
I'm still seeing this error, with pusher 1.7.2.

Stack Trace:
```
  ... 
  Application stuff
  ... 

  File "pusher/pusher.py", line 117, in trigger
    channels, event_name, data, socket_id)
  File "pusher/http.py", line 31, in __call__
    return self.client.http.send_request(self.make_request(*args, **kwargs))
  File "pusher/requests.py", line 44, in send_request
    **self.options)
  File "requests/sessions.py", line 502, in request
    resp = self.send(prep, **send_kwargs)
  File "newrelic/api/external_trace.py", line 124, in dynamic_wrapper
    return wrapped(*args, **kwargs)
  File "raven/breadcrumbs.py", line 297, in send
    resp = real_send(self, request, *args, **kwargs)
  File "requests/sessions.py", line 612, in send
    r = adapter.send(request, **kwargs)
  File "requests/adapters.py", line 490, in send
    raise ConnectionError(err, request=request)
```

**Comment by hot-leaf-juice** [comment]
@amackera What is the content of the `ConnectionError`, and what version of Python are you using? Looks like it might be related to [this issue](https://github.com/requests/requests/issues/2364), in which case we would need to do some more digging to see why the connection is being closed early. How often is this happening and do you have a reliable way to reproduce?

In the meantime you should safely be able to retry the trigger on error.

**Comment by TheKevJames** [comment]
I am also seeing this error with both: 
- python 2.7.14 (from docker image `python:2.7-slim`), and
- python 2.7.12 (from docker image `ubuntu:16.04` after an `apt-get install python`)

Relevant python libs:
```
certifi==2018.1.18
cryptography==2.1.4
pusher==1.7.2
pyOpenSSL==17.5.0
raven==6.5.0
requests==2.18.4
urllib3==1.22
```

This has occurred on several different servers running one of two potential docker images (python versions described above, python package versions are identical).

This error occurs very intermittently (in the past two days, we've seen 5 instances of this error across >100k pusher pushes), and I can not yet contribute a reliable reproduction method.

I can't share the entire stack trace for proprietary reasons, but I do have some sentry instrumentation of this error case I can share:

```
>>> pusher.trigger(['channel-1', 'channel-2'], 'event-name', ujson.dumps(pusher_payload_dictionary, double_precision=4))
```

I can confirm I've only seen this issue when we are attempting to call `trigger` with *multiple channels*, but I am not sure if this is a potential clue or just a sampling issue.

![image](https://user-images.githubusercontent.com/3341822/36817088-c719aa70-1c94-11e8-90f1-8aca9c014eb6.png)

If there's anything else I can contribute which might help debug, please let me know -- these issues are pretty well instrumented on our end, I just have to make sure to filter information out before posting it. If I can come across a reliable repro case, I will follow-up here.

**Comment by EsauPR** [comment]
I also have this error

Python 2.7
Pusher 2.0.0

The error is raised when I trigger a events throught the **trigger_batch** method.
The ratelimit is under 10 calls per second (https://pusher.com/docs/client_api_guide/client_events), I don't know if this apply for the server side.

```
('Connection aborted.', BadStatusLine("''",))
```

**Comment by WillSewell** [comment]
Reading [this thread](https://github.com/requests/requests/issues/2642), it looks like this error could be caused by a number of reasons. Are you your requests going through a proxy server? The fact it only happens with the `trigger_batch` method might mean that an HTTP proxy might not be handling larger requests properly.

Are you tracking these failure over time? E.g. graphing them? It would be interesting to know whether they happen at around the same time, or spread around evenly. If you can share this, then we (Pusher) could try and correlate these events with events on our servers (e.g. HTTP server restarts).

It could just be caused by general networking issues, out of our control. In which case, depending on your use case it might be best to retry the request.

**Comment by EsauPR** [comment]
Cause the number of messages are too many (300 messages per minute) within a very short time we use the trigger_batch method. The error is raised around every hour although the last logs show a long time between each error.

Here the question. If the number of messages to send are to many, Send the messages by http is the correct way? or Is better to create a single connection using a websocket? I am working from the server side.

Also I attach the last logs.

[pusher.log](https://github.com/pusher/pusher-http-python/files/2040276/pusher.log)
![screenshot-2018-5-25 discover - kibana](https://user-images.githubusercontent.com/5316211/40557420-b7875e8a-6015-11e8-8a9f-d93e6c55978c.png)

**Comment by WillSewell** [comment]
The fact that the errors tend to happen every hour makes me think that it's not due to Pusher HTTP server restarts because these happen much less frequently. The fact that you only see this for trigger_batch also backs this up.

> Send the messages by http is the correct way?

Yes this is the recommended approach.

Which HTTP client library backend are you using with pusher-http-python? If you are using the latest version of requests, then it should be doing HTTP keep alive. For that reason I would suggest experiment with using regular trigger (not trigger_batch) because 300 messages per minute is not extremely high.

**Comment by EsauPR** [comment]
Hi there,

We are using requests as backend library. After implementing the suggested changes (using a regular trigger) the error happens more frequently. So we decide to migrate our app to python 3 and with this change the error is diferent.

('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',))

Python and library versions:

Python docker: 3.6
pusher: 2.0.1
requests: 2.18.4

**Comment by WillSewell** [comment]
I wonder if it is caused by this issue: https://github.com/requests/requests/issues/4664

It might be worth trying to reproduce it by running this script for a while:

```
import pusher
import time

from pusher import Pusher
pusher = Pusher(app_id=u'100462', key=u'7435d7e5de69a077ebe0', secret=u'8b8400f3d5c70ae238ac', cluster=u'mt1')

while True:
    resp = pusher = Pusher(app_id=u'4', key=u'key', secret=u'secret', cluster=u'cluster')
    print(resp)
    time.sleep(0.2) # 300/minute
```

Can you work out the rate you are publishing when you experience the highest rate of these errors and try and tweak the rate the script publishes at to match it?

**Comment by WillSewell** [comment]
I wonder if it is caused by this issue: https://github.com/requests/requests/issues/4664

It might be worth trying to reproduce it by running this script for a while:

```python
import pusher
import time

from pusher import Pusher
pusher = Pusher(app_id=u'4', key=u'key', secret=u'secret', cluster=u'cluster')

while True:
    resp = pusher.trigger(u'a_channel', u'an_event', {u'some': u'data'})
    print(resp)
    time.sleep(0.2) # 300/minute
```

Can you work out the rate you are publishing when you experience the highest rate of these errors and try and tweak the rate the script publishes at to match it?

**Comment by EsauPR** [comment]
I try to reproduce the issue running the script for several hours I even removed the line for the sleep unsuccessfully. Looking in the logs for the lurllib3 versions used when the issue was present

**Comment by TheKevJames** [comment]
Hmm still seeing this error with:
```
(python 2.7.14)
pusher==1.7.2
requests==2.19.1
urllib3==1.23
```

Interestingly we **do not see this error under python 3.5.2** either under `pusher==1.7.2` or under `pusher==2.0.1`.

**Comment by WillSewell** [comment]
@TheKevJames are you able to reproduce the issue with the script I posted in https://github.com/pusher/pusher-http-python/issues/23#issuecomment-399107976?

**Comment by TheKevJames** [comment]
We've since fully migrated off of Python 2.7, so this issue is no longer relevant to us.

If I'm reading my own messages correctly (wow, this is an old ticket!) it looks like Python 3.6 support seems to have gotten resolved sometime around early July, but Python 2.7 was still broken with `pusher==1.7.2` around then; since there have been a whole bunch of `pusher` updates since then, that'l likely been resolved by now (plus, py2 is nearly EOL, so... eh, its probably fine).

**Comment by shyandsy** [comment]
it solved my problem using argument --no-sandbox

anyone explain why it is?

```
# self.browser = await pyppeteer.launch()
self.browser = await pyppeteer.launch(options={'args': ['--no-sandbox']})

```
